File: static/js/wiki-editor/utils/namespace.js
// File: static/js/wiki-editor/utils/namespace.js
/**
 * Namespace utilities for the wiki editor
 */

/**
 * Valid namespaces configuration
 */
export const NAMESPACES = {
    '': {
        name: 'Main',
        description: 'Main content namespace for articles',
        urlPrefix: '/articles',
        searchable: true,
        allowCategories: true
    },
    'Category': {
        name: 'Category',
        description: 'Category organization pages',
        urlPrefix: '/categories',
        searchable: true,
        allowCategories: false
    },
    'Template': {
        name: 'Template',
        description: 'Reusable template content',
        urlPrefix: '/templates',
        searchable: false,
        allowCategories: true
    },
    'Help': {
        name: 'Help',
        description: 'Help and documentation pages',
        urlPrefix: '/help',
        searchable: true,
        allowCategories: true
    },
    'User': {
        name: 'User',
        description: 'User pages and subpages',
        urlPrefix: '/users',
        searchable: false,
        allowCategories: false
    },
    'File': {
        name: 'File',
        description: 'File description pages',
        urlPrefix: '/media',
        searchable: true,
        allowCategories: true
    },
    'Kryptopedia': {
        name: 'Kryptopedia',
        description: 'Project-related pages',
        urlPrefix: '/project',
        searchable: true,
        allowCategories: true
    },
    'Talk': {
        name: 'Talk',
        description: 'Discussion pages',
        urlPrefix: '/talk',
        searchable: false,
        allowCategories: false
    }
};

/**
 * Parse namespace from a title
 * @param {string} title - Full title with potential namespace
 * @returns {Object} Object with namespace and title properties
 */
export function parseNamespaceTitle(title) {
    if (title.includes(':')) {
        const [potentialNamespace, ...titleParts] = title.split(':');
        
        if (NAMESPACES[potentialNamespace]) {
            return {
                namespace: potentialNamespace,
                title: titleParts.join(':').trim(),
                fullTitle: title
            };
        }
    }
    
    // No valid namespace found
    return {
        namespace: '',
        title: title.trim(),
        fullTitle: title
    };
}

/**
 * Format title with namespace
 * @param {string} namespace - The namespace (empty for main)
 * @param {string} title - The title
 * @returns {string} Formatted full title
 */
export function formatNamespaceTitle(namespace, title) {
    if (namespace) {
        return `${namespace}:${title}`;
    }
    return title;
}

/**
 * Get URL for a namespace and title
 * @param {string} namespace - The namespace
 * @param {string} title - The title
 * @returns {string} URL path
 */
export function getNamespaceUrl(namespace, title) {
    const urlTitle = title.replace(/ /g, '_');
    const config = NAMESPACES[namespace] || NAMESPACES[''];
    return `${config.urlPrefix}/${urlTitle}`;
}

/**
 * Check if namespace is valid
 * @param {string} namespace - Namespace to check
 * @returns {boolean} True if valid
 */
export function isValidNamespace(namespace) {
    return namespace in NAMESPACES;
}

/**
 * Get namespace display name
 * @param {string} namespace - Namespace key
 * @returns {string} Display name
 */
export function getNamespaceDisplayName(namespace) {
    const config = NAMESPACES[namespace];
    return config ? config.name : 'Unknown';
}

/**
 * Check if namespace allows categories
 * @param {string} namespace - Namespace to check
 * @returns {boolean} True if categories are allowed
 */
export function namespaceAllowsCategories(namespace) {
    const config = NAMESPACES[namespace];
    return config ? config.allowCategories : false;
}

/**
 * Get searchable namespaces
 * @returns {Array} Array of searchable namespace keys
 */
export function getSearchableNamespaces() {
    return Object.keys(NAMESPACES).filter(ns => NAMESPACES[ns].searchable);
}

/**
 * Extract namespace references from content
 * @param {string} content - Wiki markup content
 * @returns {Array} Array of namespace references
 */
export function extractNamespaces(content) {
    const namespaces = new Set();
    
    // Find all internal links that might have namespaces
    const linkPattern = /\[\[([^\]|]+)(?:\|[^\]]*)?\]\]/g;
    let match;
    
    while ((match = linkPattern.exec(content)) !== null) {
        const linkTarget = match[1];
        const parsed = parseNamespaceTitle(linkTarget);
        
        if (parsed.namespace) {
            namespaces.add(parsed.namespace);
        }
    }
    
    return Array.from(namespaces);
}

/**
 * Suggest appropriate namespace based on title patterns
 * @param {string} title - The page title
 * @returns {string} Suggested namespace (empty string for main)
 */
export function suggestNamespaceForTitle(title) {
    const titleLower = title.toLowerCase();
    
    // Category pattern
    if (titleLower.startsWith('category')) {
        return 'Category';
    }
    
    // Template pattern  
    if (titleLower.startsWith('template')) {
        return 'Template';
    }
        
    // Help pattern
    if (['help', 'guide', 'tutorial', 'how to'].some(word => titleLower.includes(word))) {
        return 'Help';
    }
        
    // User pattern
    if (titleLower.startsWith('user')) {
        return 'User';
    }
        
    // File pattern
    if (['.jpg', '.png', '.gif', '.svg', '.pdf', '.doc'].some(ext => titleLower.includes(ext))) {
        return 'File';
    }
        
    // Project pattern
    if (['kryptopedia', 'policy', 'guideline', 'project'].some(word => titleLower.includes(word))) {
        return 'Kryptopedia';
    }
        
    // Default to main namespace
    return '';
}

/**
 * Validate namespace for a title
 * @param {string} namespace - Namespace to validate
 * @param {string} title - Title to validate
 * @returns {Object} Validation result
 */
export function validateNamespace(namespace, title) {
    const result = {
        valid: true,
        warnings: [],
        errors: []
    };
    
    // Check if namespace exists
    if (namespace && !isValidNamespace(namespace)) {
        result.valid = false;
        result.errors.push(`Invalid namespace: ${namespace}`);
        return result;
    }
    
    // Check namespace-specific rules
    const config = NAMESPACES[namespace] || NAMESPACES[''];
    
    // Category namespace should have category-like titles
    if (namespace === 'Category' && !title.match(/^[A-Z]/)) {
        result.warnings.push('Category titles should start with a capital letter');
    }
    
    // Template namespace recommendations
    if (namespace === 'Template' && title.includes(' ')) {
        result.warnings.push('Template names should avoid spaces when possible');
    }
    
    // File namespace should have file extensions
    if (namespace === 'File' && !title.match(/\.[a-zA-Z0-9]+$/)) {
        result.warnings.push('File pages should include file extensions');
    }
    
    return result;
}

File: utils/namespace.py
# File: utils/namespace.py
"""
Namespace utilities for Kryptopedia.
"""
from typing import Dict, List, Optional, Tuple

# Define the namespace configuration
NAMESPACE_CONFIG = {
    "": {
        "name": "Main",
        "description": "Main content namespace for articles",
        "url_prefix": "/articles",
        "searchable": True,
        "allow_categories": True,
        "content_model": "wikitext"
    },
    "Category": {
        "name": "Category", 
        "description": "Category organization pages",
        "url_prefix": "/categories",
        "searchable": True,
        "allow_categories": False,
        "content_model": "category"
    },
    "Template": {
        "name": "Template",
        "description": "Reusable template content",
        "url_prefix": "/templates", 
        "searchable": False,
        "allow_categories": True,
        "content_model": "wikitext"
    },
    "Help": {
        "name": "Help",
        "description": "Help and documentation pages",
        "url_prefix": "/help",
        "searchable": True,
        "allow_categories": True,
        "content_model": "wikitext"
    },
    "User": {
        "name": "User",
        "description": "User pages and subpages",
        "url_prefix": "/users",
        "searchable": False,
        "allow_categories": False,
        "content_model": "wikitext"
    },
    "File": {
        "name": "File", 
        "description": "File description pages",
        "url_prefix": "/media",
        "searchable": True,
        "allow_categories": True,
        "content_model": "file"
    },
    "Kryptopedia": {
        "name": "Kryptopedia",
        "description": "Project-related pages",
        "url_prefix": "/project",
        "searchable": True,
        "allow_categories": True,
        "content_model": "wikitext"
    },
    "Talk": {
        "name": "Talk",
        "description": "Discussion pages",
        "url_prefix": "/talk",
        "searchable": False,
        "allow_categories": False,
        "content_model": "wikitext"
    }
}

def get_valid_namespaces() -> Dict[str, str]:
    """Get a mapping of namespace keys to display names."""
    return {ns: config["name"] for ns, config in NAMESPACE_CONFIG.items()}

def is_valid_namespace(namespace: str) -> bool:
    """Check if a namespace is valid."""
    return namespace in NAMESPACE_CONFIG

def get_namespace_info(namespace: str) -> Optional[Dict]:
    """Get configuration information for a namespace."""
    return NAMESPACE_CONFIG.get(namespace)

def parse_title_with_namespace(full_title: str) -> Tuple[str, str]:
    """
    Parse a title to extract namespace and title.
    
    Args:
        full_title: The full title (e.g., "Category:Blockchain" or "Introduction to Crypto")
        
    Returns:
        Tuple[str, str]: (namespace, title) - namespace is empty string for main namespace
    """
    if ":" in full_title:
        potential_namespace, title = full_title.split(":", 1)
        if is_valid_namespace(potential_namespace):
            return potential_namespace, title.strip()
    
    # No valid namespace found, treat as main namespace
    return "", full_title.strip()

def get_namespace_url(namespace: str, title: str) -> str:
    """
    Generate the appropriate URL for a page in a namespace.
    All articles stay under /articles/ path regardless of namespace.
    
    Args:
        namespace: The namespace
        title: The page title
        
    Returns:
        str: The URL path
        
    Examples:
        get_namespace_url("Kryptopedia", "Rules (being merged)")
        → "/articles/Kryptopedia:Rules_(being_merged)"
        
        get_namespace_url("", "Bitcoin Basics") 
        → "/articles/Bitcoin_Basics"
    """
    from .slug import generate_namespace_slug
    
    # Generate the proper slug that includes namespace
    slug = generate_namespace_slug(namespace, title)
    
    # All articles use the same /articles/ path
    return f"/articles/{slug}"

def format_full_title(namespace: str, title: str) -> str:
    """
    Format a full title including namespace prefix.
    
    Args:
        namespace: The namespace (empty string for main)
        title: The page title
        
    Returns:
        str: The full formatted title
    """
    if namespace:
        return f"{namespace}:{title}"
    return title

def get_searchable_namespaces() -> List[str]:
    """Get list of namespaces that should be included in search."""
    return [ns for ns, config in NAMESPACE_CONFIG.items() if config["searchable"]]

def namespace_allows_categories(namespace: str) -> bool:
    """Check if a namespace allows category assignment."""
    config = get_namespace_info(namespace)
    return config["allow_categories"] if config else False

def get_namespace_content_model(namespace: str) -> str:
    """Get the content model for a namespace."""
    config = get_namespace_info(namespace)
    return config["content_model"] if config else "wikitext"

def suggest_namespace_for_title(title: str) -> str:
    """
    Suggest appropriate namespace based on title patterns.
    
    Args:
        title: The page title
        
    Returns:
        str: Suggested namespace (empty string for main)
    """
    title_lower = title.lower()
    
    # Category pattern
    if title_lower.startswith("category"):
        return "Category"
    
    # Template pattern  
    if title_lower.startswith("template"):
        return "Template"
        
    # Help pattern
    if any(word in title_lower for word in ["help", "guide", "tutorial", "how to"]):
        return "Help"
        
    # User pattern
    if title_lower.startswith("user"):
        return "User"
        
    # File pattern
    if any(ext in title_lower for ext in [".jpg", ".png", ".gif", ".svg", ".pdf", ".doc"]):
        return "File"
        
    # Project pattern
    if any(word in title_lower for word in ["kryptopedia", "policy", "guideline", "project"]):
        return "Kryptopedia"
        
    # Default to main namespace
    return ""

def get_namespace_statistics() -> Dict[str, Dict[str, int]]:
    """
    Get statistics about namespace usage.
    This would typically query the database, but returns empty for now.
    
    Returns:
        Dict[str, Dict[str, int]]: Statistics per namespace
    """
    stats = {}
    for namespace in NAMESPACE_CONFIG.keys():
        stats[namespace] = {
            "total_pages": 0,
            "total_views": 0,
            "recent_changes": 0
        }
    return stats
